src/Features/LanguageServer/Protocol/Extensions/Extensions.cs

```diff
@@ -78,16 +76,25 @@ public static ImmutableArray<TextDocument> GetTextDocuments(this Solution soluti
             var documentIds = GetDocumentIds(solution, documentUri);
 
             var documents = documentIds
-                .Select(solution.GetDocument)
-                .Concat(documentIds.Select(solution.GetAdditionalDocument))
+                .Select(solution.GetTextDocument)
                 .WhereNotNull()
                 .ToImmutableArray();
             return documents;
         }
 
         public static ImmutableArray<DocumentId> GetDocumentIds(this Solution solution, Uri documentUri)
-            => solution.GetDocumentIdsWithFilePath(ProtocolConversions.GetDocumentFilePathFromUri(documentUri));
+        {
+            // If this is not our special scheme for generated documents, then we can just look for documents with that file path.
+            if (documentUri.Scheme != SourceGeneratedDocumentUris.Scheme)
+                return solution.GetDocumentIdsWithFilePath(ProtocolConversions.GetDocumentFilePathFromUri(documentUri));
+
+            var documentId = SourceGeneratedDocumentUris.DeserializeIdentity(solution, documentUri)?.DocumentId;
```

Comment:
we were given a source generated document identity, should we throw if we can't deserialize?

Reply:
throwing seems good :)

Reply:
We'll throw if we can't deserialize; the case where we won't throw but return null is if we are being given a document ID that we know is stale, say because the project has been unloaded. I can make that throw too if we want, but that might hit the user.

Reply:
(and the project ID check is mostly so we can recapture the "real" project ID for purposes of debugging)

Reply:
Hmm.  The unloaded case I assume would be better handled once we are able to refresh the document.  But even if we tell the client it is gone, it is possible the client already has a request in flight by the time it gets told it is removed.

In that case returning null for a documentId seems reasonable - since the document was not found.

Reply:
> it is possible the client already has a request in flight by the time it gets told it is removed.

Yep you're correctly thinking about the potential race here.